git_buildbot work for python3

dannyvi opened the pull request:
Add a git_buildbot_py3.py file works for post-receive script calling.

dannyvi committed:
git_buildbot work for python3

oliver-joos commented:
Thanks dannyvi! I tested your "git_buildbot_py3.py" against Buildbot 1.0.0 / Twisted 17.9.0 / Python 3.5.2.
**It works like a charm!** The original "git_buildbot.py" only works with Python 2 and (even worse) only if the Buildbot master also runs on Python 2 (on Python 3 changedict gets keys of type bytes and then **changedict fails)

My findings:
1. Your script also works with Python 2 (2.7.12) so I propose to replace current "git_buildbot.py" with yours.
2. Your script uses a different default port 9999. This may hurt beginners. You should change it to 9989.
3. python3-twisted 16.0.0 fails to import twisted.spread.pb. Unfortunately 16.0.0 is still wide-spread in latest Ubuntu 16.04.3 LTS and Mint 18.3. It may be helpful to add a comment or even a log message upon ImportError. I fixed this problem by upgrading to Twisted 17.9.0 with terminal command `sudo -H pip3 install Twisted`

dannyvi committed:
moved git_buildbot.py, corrected port 9989, added comment to ImportError

dannyvi commented:
oliver-joos, thanks for your reply and the test works.
Follow your instructions, I did some work in the new commit.

   1. Move the "git_buildbot_py3.py" to "git_buildbot.py", and deleted the origin file.
   2. Corrected the port to 9989.
   3. Added a comment to the ImportError for twisted.spread.pb.

Hope this would perform a good work, and any other suggests would be welcomed.

tardyp commented on master/contrib/git_buildbot.py:
```diff
@@ -208,11 +224,17 @@ def gen_create_branch_changes(newrev, refname, branch):
     logging.info("Branch `%s' created", branch)
 
-    f = os.popen("git rev-parse --not --branches"
+    #f = os.popen("git rev-parse --not --branches"
```
we should avoid commented code

dannyvi replied:
Sorry to forgot doing that.

dannyvi committed:
removed code comments

dannyvi committed:
bugfix remove the last os.popen

hraphrap commented:
Hi, @dannyvi! Thanks for you job!
I have Python version 3.4.8 and git version 1.7.10.4 on my Gitolite server and i installed git_buildbot.py from your fork of buildbot-contrib repository. It works well with existing branches, but when i create new branch it shows me an error:
`remote: git_buildbot: INFO: Branch `testing_branch' created
remote: fatal: ambiguous argument 'grep': unknown revision or path not in the working tree.
remote: Use '--' to separate paths from revisions
remote: git_buildbot: DEBUG: Change: --branches|
remote:
remote: git_buildbot: ERROR: Unhandled exception
remote: Traceback (most recent call last):
remote:   File "hooks/post-receive.h00-git_buildbot.py", line 487, in <module>
remote:     process_changes()
remote:   File "hooks/post-receive.h00-git_buildbot.py", line 374, in process_changes
remote:     process_change(oldrev, newrev, refname)
remote:   File "hooks/post-receive.h00-git_buildbot.py", line 358, in process_change
remote:     process_branch_change(oldrev, newrev, refname, branch)
remote:   File "hooks/post-receive.h00-git_buildbot.py", line 336, in process_branch_change
remote:     gen_create_branch_changes(newrev, refname, branch)
remote:   File "hooks/post-receive.h00-git_buildbot.py", line 231, in gen_create_branch_changes
remote:     gen_changes(f, branch)
remote:   File "hooks/post-receive.h00-git_buildbot.py", line 195, in gen_changes
remote:     c = {'revision': m.group(1),
remote: AttributeError: 'NoneType' object has no attribute 'group'
`

oliver-joos commented:
I just found another bug caused by non-ascii characters in commit comments.
The causing comment has 3 lines (2nd empty):
```
    Dokumentation
    
    Add Module Übersicht
```

The traceback is here:
(I only replaced personal infos like email, author and branch names with "xyz")
```
2018-05-08 15:11:26,934 ERROR: 3.5.2 (default, Nov 23 2017, 16:37:01)
[GCC 5.4.0 20160609]
2018-05-08 15:11:26,936 INFO: Branch `xyz' updated 3207ff01 .. e8c1b746
2018-05-08 15:11:26,947 DEBUG: oldrev=3207ff015dd7db798f340f58cb7b4d49225aaa46 newrev=e8c1b746b5091bf1db2db3dcfce7f9cac82bfa6b baserev=3207ff015
2018-05-08 15:11:26,952 DEBUG: Change: e8c1b746b5091bf1db2db3dcfce7f9cac82bfa6b Dokumentation

2018-05-08 15:11:26,961 DEBUG: Got author: Xyz <xyz@xyz.ch>
2018-05-08 15:11:26,961 ERROR: Unhandled exception
Traceback (most recent call last):
  File "./hooks/post-receive-buildbot", line 494, in <module>
    process_changes()
  File "./hooks/post-receive-buildbot", line 379, in process_changes
    process_change(oldrev, newrev, refname)
  File "./hooks/post-receive-buildbot", line 361, in process_change
    process_branch_change(oldrev, newrev, refname, branch)
  File "./hooks/post-receive-buildbot", line 341, in process_branch_change
    gen_update_branch_changes(oldrev, newrev, refname, branch)
  File "./hooks/post-receive-buildbot", line 323, in gen_update_branch_changes
    gen_changes(f, branch)
  File "./hooks/post-receive-buildbot", line 209, in gen_changes
    grab_commit_info(c, m.group(1))
  File "./hooks/post-receive-buildbot", line 155, in grab_commit_info
    line = f.stdout.readline().decode("ascii")
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 15: ordinal not in range(128)
```

oliver-joos commented:
I found a quick solution, at least for our Buildbot server!

There is a global variable named "encoding" which is set by option "--encoding". But this variable is never used! It was already set to 'utf8' on our server. But nevertheless there are 4 code lines that use 'ascii' to decode stdout from calls of shell-commands like "git".

I replaced all 4 of them (line 155, 186, 271 and 286) like this:
`...stdout.readline().decode("ascii")`  =>  `...stdout.readline().decode(encoding)`

Because 'ascii' is a subset of 'utf8' I am pretty sure that my fix won't break existing setups. But the Buildbot server has to handle such utf8-encoded strings correctly! We use Buildbot 1.1.1 on Twisted 17.9.0 and now non-ascii characters do trigger a build and are also displayed correctly (e.g. in Console View). And second, I did not understand the use of global variable "encoding" and hijacked it for my purpose. I am not sure if this was a wise move.

Please consider both of my concerns and make some own tests before you merge this fix into your pull request.

oliver-joos commented:
We use my fix now for almost a month in different projects, and I think it is safe to be merged.

@dannyvi: what do you think about my fix? And do you know, why your git_buildbot.py has not been merged upstream yet?

oliver-joos commented:
This pull request has been merged, so I moved the fix to https://github.com/buildbot/buildbot-contrib/pull/13

mrstanwell commented:
@hraphrap : I had the same problem.  I've submitted a pull request that fixes it for me.  See #15
